File: 153.find-minimum-in-rotated-sorted-array.js
/*
 * @lc app=leetcode id=153 lang=javascript
 *
 * [153] Find Minimum in Rotated Sorted Array
 */
/**
 * @param {number[]} nums
 * @return {number}
 */
var findMin = function(nums) {
	// smallest now is nums[end];
	// keep the feature for 拐点
	let l = 0,
		r = nums.length - 1;

	while (l + 1 < r) {
		let mid = Math.floor((l + r) / 2);
		if (nums[mid] < nums[r]) {
			r = mid;
		} else {
			l = mid;
		}
	}

	return Math.min(nums[l], nums[r]);
};

findMin([ 3, 4, 5, 1, 2 ]);

File: unsorted/java/153.find-minimum-in-rotated-sorted-array.java
/*
 * @lc app=leetcode id=153 lang=java
 *
 * [153] Find Minimum in Rotated Sorted Array
 */

// @lc code=start
class Solution {
    public int findMin(int[] nums) {
        // [l...r]
        int l=0, r=nums.length-1;
        while(l+1<r){
            int m = l + (r-l)/2;
            if(nums[m] < nums[nums.length-1]){
                r = m;
            }else{
                l = m;
            }
        }
        return Math.min(nums[r], nums[l]);
        
    }
}
// @lc code=end

